### PR TITLE
Update user-server and e2e

### PR DESCRIPTION
**Pull request by xuezhaojun**
Signed-off-by: xuezhaojun <zxue@redhat.com>

### CODE REVIEW

File: cmd/cluster-proxy/main.go [hunk 1]
```diff
@@ -52,11 +50,12 @@ func newClusterProxyCommand() *cobra.Command {
 		cmd.Version = v
 	}
 
-	cmd.AddCommand(hub.NewController())
 	cmd.AddCommand(hub.NewUserServer())
-	cmd.AddCommand(spoke.NewAgent())
-	cmd.AddCommand(spoke.NewAPIServerProxy())
-	cmd.AddCommand(configchecker.NewConfigCheckerServer())
+	// Using cluster-proxy as the base, so no needs for the following controllers.
+	// cmd.AddCommand(hub.NewController())
```
Comment (qiujian16) [inline cmd/cluster-proxy/main.go:55]: add a TODO to remove them

File: pkg/cmd/hub/user_proxy.go [hunk 2]
```diff
@@ -1,24 +1,23 @@
...
-
 	"github.com/openshift/library-go/pkg/controller/controllercmd"
 	"github.com/spf13/cobra"
+	"github.com/stolostron/cluster-proxy-addon/pkg/hub"
 	"github.com/stolostron/cluster-proxy-addon/pkg/version"
 )
 
 func NewUserServer() *cobra.Command {
-	userServerOptions := hub.NewUserServerOptions()
+	kubectlUI := hub.NewKubectlUI()
```
Comment (qiujian16) [inline pkg/cmd/hub/user_proxy.go:11]: nit, I would call it HTTPServer or HTTPUserServer. It is not only used by kubectl

File: pkg/hub/kubectl_ui.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true, // Skip server-auth for kube-apiserver
+		},
+		// golang http pkg automaticly upgrade http connection to http2 connection, but http2 can not upgrade to SPDY which used in "kubectl exec".
+		// set ForceAttemptHTTP2 = false to prevent auto http2 upgration
+		ForceAttemptHTTP2: false,
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			proxyConn, err = tunnel.DialContext(ctx, network, addr)
```
Comment (qiujian16) [inline pkg/hub/kubectl_ui.go:144]: I wonder if we should maintain a long running connection here. For instance, you could have a controller that establish the connection based on the status of addon resources, and put it in the cache. Will that enhance performance?

**Review by qiujian16 — COMMENTED** [review]
/approve

I think in general it looks good. I have some nit comment and a question.

@skeeey PTAL

File: pkg/hub/kubectl_ui.go (hunk 3 above)
Comment (xuezhaojun) [reply pkg/hub/kubectl_ui.go:144]: Yes, this is also a part of the next plan.

File: cmd/cluster-proxy/main.go (hunk 1 above)
Comment (xuezhaojun) [reply cmd/cluster-proxy/main.go:55]: Fixed.

File: pkg/cmd/hub/user_proxy.go (hunk 2 above)
Comment (xuezhaojun) [reply pkg/cmd/hub/user_proxy.go]: Name changed.

File: pkg/hub/http_user_server.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	target := fmt.Sprintf("https://%s", clusterID)
+	apiserverURL, err := url.Parse(target)
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	tunnel, err := k.getTunnel()
```
Comment (skeeey) [inline pkg/hub/http_user_server.go:114]: the tunnel seems to be same for each request, right? if it's right, why we create the tunnel every time?

File: pkg/hub/http_user_server.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true, // Skip server-auth for kube-apiserver
+		},
+		// golang http pkg automaticly upgrade http connection to http2 connection, but http2 can not upgrade to SPDY which used in "kubectl exec".
+		// set ForceAttemptHTTP2 = false to prevent auto http2 upgration
+		ForceAttemptHTTP2: false,
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			proxyConn, err = tunnel.DialContext(ctx, network, addr)
```
Comment (skeeey) [inline pkg/hub/http_user_server.go:144]: you may need a TODO here as a reminder

Comment (xuezhaojun) [reply pkg/hub/http_user_server.go:144]: It's not a reminder comment but a descriptional comment.

File: pkg/hub/http_user_server.go (hunk 4 above)
Comment (xuezhaojun) [reply pkg/hub/http_user_server.go:114]: In our test, the tunnel is not supported reused many times. This should be an enhancement TODO in the next step.

File: pkg/hub/http_user_server.go (hunk 5 above)
Comment (skeeey) [reply pkg/hub/http_user_server.go:144]: I mean the comments for cache the proxyConn :)

File: pkg/hub/http_user_server.go (hunk 4 above)
Comment (skeeey) [reply pkg/hub/http_user_server.go:114]: Add a comments here?

Comment (xuezhaojun) [reply pkg/hub/http_user_server.go:114]: Sure, the TODO comment added.

File: pkg/hub/http_user_server.go (hunk 5 above)
Comment (xuezhaojun) [reply pkg/hub/http_user_server.go:144]: I'm not sure if the proxyConn should be cached and reused as the tunnel. But I will set a TODO here.

**Pull request merged**